flightControl/src/AvState.cpp

```diff
@@ -0,0 +1,262 @@
+#include "AvState.h"
+#include <string.h>
+#include <list>
+#include <sensors.h>
+#include "telecom.h"
+#include "PacketDefinition.h"
+
+
+AvState::AvState(const Thresholds& thresholds) : thresholds(thresholds)
+{
+    this->currentState = State::IDLE;
+}
+
+// destructor
+AvState::~AvState()
+{
+    currentState = State::IDLE;
+}
+
+
+// this function allows to get the current state of the FSM
+State AvState::getCurrentState()
+{
+    return currentState;
+}
+
+
+void sendError()
+{
+    //print in console the "error"
+    std::cout << "Error" << std::endl;
+}
+
+bool error()
+{
+    return false;
+}
+
+
+State AvState::fromIdle(SensFiltered data, UPLink uplink)
+{
+    if ((uplink.id == CMD_ID::AV_CMD_CALIBRATE))
+    {
+        return State::CALIBRATION;
+    }
+    return State::IDLE;
+}
+
+State AvState::fromLanded() { return State::LANDED; }
+
+State AvState::fromDescent(SensFiltered data, UPLink uplink)
+{
+    //norm of the speed vector
+    if (uplink.id == CMD_ID::AV_CMD_ABORT || uplink.id == CMD_ID::AV_CMD_MANUAL_DEPLOY)
+    {
+        return State::ERRORFLIGHT;
+    }
+    double speed = (data.speed.x * data.speed.x + data.speed.y * data.speed.y + data.speed.z * data.speed.z);
+    if (speed < 0)
+    {
+        return State::LANDED;
+    }
+    return State::DESCENT;
+}
+
+State AvState::fromAscent(SensFiltered data, UPLink uplink)
+{
+    if (uplink.id == CMD_ID::AV_CMD_ABORT)
+    {
+        return State::ERRORFLIGHT;
+    }
+    else if (data.speed.z <= 0)
+    {
+        return State::DESCENT;
+    }
+    return State::ASCENT;
+}
+
+State AvState::fromCalibration(SensFiltered data, UPLink uplink, bool status)
+{
+    if (error() || uplink.id == CMD_ID::AV_CMD_ABORT)
+    {
+        return State::ERRORGROUND;
+        //check if the calibration is done
+    }
+    else if (status)
+    {
+        return State::MANUAL;
+    }
+    else
+    {
+        return State::CALIBRATION;
+    }
+}
+
+
+State AvState::fromErrorGround(SensFiltered data, UPLink uplink)
+{
+    if (uplink.id == CMD_ID::AV_CMD_RECOVER)
+    {
+        return State::IDLE;
+    }
+}
+
+State AvState::fromErrorFlight()
+{
+    return State::ERRORFLIGHT;
+}
+
+State AvState::fromThrustSequence(SensFiltered data, UPLink uplink)
+{
+    double speed = (data.speed.x * data.speed.x + data.speed.y * data.speed.y + data.speed.z * data.speed.z);
+    if (uplink.id == CMD_ID::AV_CMD_ABORT)
+    {
+        return State::ERRORFLIGHT;
+    }
+    else if (data.speed.z>0) //TODO we need a ignite bool var
+                                //given by the prop board
+    {
+        return State::ASCENT;
+    }
+    else
+    {
+        return State::THRUSTSEQUENCE;
+    }
+}
+
+State AvState::fromManual(SensFiltered data, UPLink uplink)
+{
+    if (error() || uplink.id == CMD_ID::AV_CMD_ABORT)
+    {
+        return State::ERRORGROUND;
+    }
+        //TODO: recheck i threholds are the wanted ones
+    else if (uplink.id == CMD_ID::AV_CMD_ARM &&
```

Comment:
This must be rewritten. The transition could be check elsewise. A happy-path alternative would allow for precise logging and feedback.

```suggestion
    if (uplink.id != CMD_ID::AV_CMD_ARM)
    {
        // No ARM order is received, thus no need to log anything.
        return State::MANUAL;
    }
    
    if (data.something < threshold)
    {
        // The ARM order is discraded because a certain threshold was not met. This is logged and fed back to ground control.
        return State::MANUAL;
    }
    
    // Do this for each threshold.
```

Reply:
C'est pas un problème. Cette suggestion ne concerne que le "format" du code, pas les thresholds.